### PR TITLE
Improve batch_norm backward performance for automatically generated backward

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Base PR https://github.com/Lightning-AI/lightning-thunder/pull/139/ (this PR will remain in draft mode before merging 139)

This PR reorders some operations to make disabling bookend optimization feasible without hitting a bug in nvFuser (https://github.com/NVIDIA/Fuser/issues/1964).

`nv_enable_bookend` is set to True by default. nvFuser generates a single kernel for batch norm backward when the bookend optimization is turned off. To force the nvFuser executor to skip using this optimization I inserted a redundant dtype conversion in the var_mean backward before expanding mean.
Before this change (on https://github.com/Lightning-AI/lightning-thunder/pull/139/), 3 nvFuser kernels:
```bash
nsys nvprof pytest thunder/benchmarks/targets.py -k "test_batch_norm_grad[thunder]"

Executing 'cuda_gpu_kern_sum' stats report

 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     28.6        2,655,400        220  12,070.0  12,064.0    11,905    13,952        146.0  <unnamed>::nvfuser_reduction_f1_c1_r0_g1(<unnamed>::Tensor<<unnamed>::__bfloat, (int)3, (int)3>, <u…
     21.7        2,015,528      1,100   1,832.3   1,248.0     1,184     4,257      1,186.2  void at::native::<unnamed>::distribution_elementwise_grid_stride_kernel<float, (int)4, void at::nat…
     19.0        1,767,058        220   8,032.1   8,032.0     7,936     8,192         40.6  <unnamed>::nvfuser_pointwise_f1_c1_r0_g0(<unnamed>::Tensor<<unnamed>::__bfloat, (int)3, (int)3>, <u…
     16.6        1,540,891        220   7,004.1   7,008.0     6,817     7,520         66.1  <unnamed>::nvfuser_inner_persistent_f0_c1_r0_g0(<unnamed>::Tensor<<unnamed>::__bfloat, (int)3, (int…
      8.1          749,127        880     851.3     864.0       800       896         17.9  void at::native::vectorized_elementwise_kernel<(int)4, at::native::FillFunctor<long>, at::detail::A…
      6.0          560,484        220   2,547.7   2,560.0     2,527     2,592         17.1  void at::native::vectorized_elementwise_kernel<(int)4, at::native::FillFunctor<c10::BFloat16>, at::…

```

Current PR (2 nvFuser kernels, one for forward, one for backward):
```bash
nsys nvprof pytest thunder/benchmarks/targets.py -k "test_batch_norm_grad[thunder]"

Executing 'cuda_gpu_kern_sum' stats report

 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     40.5        2,008,029      1,100   1,825.5   1,248.0     1,215     4,256      1,174.6  void at::native::<unnamed>::distribution_elementwise_grid_stride_kernel<float, (int)4, void at::nat…
     26.9        1,334,931        220   6,067.9   6,049.0     5,984     6,368         33.7  <unnamed>::nvfuser_inner_persistent_f1_c1_r0_g0(<unnamed>::Tensor<<unnamed>::__bfloat, (int)3, (int…
     21.4        1,063,366        220   4,833.5   4,832.0     4,735     5,216         66.2  <unnamed>::nvfuser_inner_persistent_f0_c1_r0_g0(<unnamed>::Tensor<<unnamed>::__bfloat, (int)3, (int…
     11.2          557,249        220   2,533.0   2,528.0     2,496     2,592         12.8  void at::native::vectorized_elementwise_kernel<(int)4, at::native::FillFunctor<c10::BFloat16>, at::…
```
Main before https://github.com/Lightning-AI/lightning-thunder/pull/139/ has the same 2 nvFuser kernels, one for forward, one for backward.

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -2513,20 +2513,20 @@ def _native_batch_norm(
     a_acc = to(a, computation_dtype)
     if training:
         reduction_dims = (0,) + tuple(range(2, a.ndim))
-        biased_var, mean = var_mean(a_acc, dim=reduction_dims, correction=0, keepdim=True)
+        biased_var, mean = var_mean(a_acc, dim=reduction_dims, correction=0, keepdim=False)
```
Comment (IvanYashchuk) [inline thunder/torch/__init__.py:2516]: This change allows for disabling the bookend optimization without hitting a bug in nvFuser (https://github.com/NVIDIA/Fuser/issues/1964).

Comment (jjsjann123) [reply thunder/torch/__init__.py:2516]: Thanks for catching this!

**Review by jjsjann123 — APPROVED** [review]
:clap:

**Comment by jjsjann123** [comment]
> Base PR https://github.com/Lightning-AI/lightning-thunder/pull/139 (this PR will remain in draft mode before merging 139)

Should this PR be merged into #139 first before we land that? Just so we are not having a `bad commit` with perf regression.

**Comment by IvanYashchuk** [comment]
We should embrace small diffs and stacked commits! 

But we can also merge this PR into the preceding one first if it makes you worry less.

**Comment by IvanYashchuk** [comment]
GitHub behaves weirdly with automerge enabled and stacked PRs.

**Review by kiya00 — APPROVED** [review]
Thank you very much @IvanYashchuk

File: thunder/torch/__init__.py (hunk 1 above)
Comment (t-vi) [reply thunder/torch/__init__.py:2530]:
```suggestion
        # this should be keepdim=False  because of https://github.com/NVIDIA/Fuser/issues/1964
        biased_var, mean = var_mean(a_acc, dim=reduction_dims, correction=0, keepdim=False)
```

**Review by t-vi — APPROVED** [review]
Thank you @IvanYashchuk @kiya00 @jjsjann123
